### PR TITLE
Add power consumption metric to power supply monitoring module

### PR DESCRIPTION
**Pull request by eyusupov**
##### Summary
Add a new metric for battery discharge rate (`/sys/class/battery/<BAT>/power_now`).
image

##### Test Plan
- Single battery when charging
- Single battery when discharging

##### Additional Information
There is a new metric that allows to track the machine (usually laptop) power consumption.
While it is possible to gauge this using existing power capacity metric, it is not very granular does not show the rate of discharge very clearly.
With this change it should possible to correlate other metrics with power consumption.

There seem to be some issues with the sections (BAT0 has appeared as the section heading and `capacity` metric is displayed under `charge` section). I think the reason might be in the dashboard code which source is not included in the repo.

### CODE REVIEW

**Comment by ilyam8** [comment]
Hey, @eyusupov 👋 Thanks for the contribution!


> There seem to be some issues with the sections

The subsection name is the 4th parameter in `rrdset_create_localhost()`. I fixed them in #17338.

**Comment by ilyam8** [comment]
@eyusupov I merged #17338, please 

- Rebase/resolve merge conflicts.
- Update the screenshot in the PR description.

**Comment by eyusupov** [comment]
I rebased my branch and updated the section (and priority) for the new metric, but it seems like it has stopped displaying completely now.

**Comment by eyusupov** [comment]
image

File: src/collectors/proc.plugin/sys_class_power_supply.c [hunk 1]
```diff
@@ -116,14 +121,82 @@ static void add_labels_to_power_supply(struct power_supply *ps, RRDSET *st) {
...
+    if(unlikely(!prop->st)) {
+        char id[RRD_ID_LENGTH_MAX + 1], context[RRD_ID_LENGTH_MAX + 1];
+        snprintfz(id, RRD_ID_LENGTH_MAX, "powersupply_%s", dim);
+        snprintfz(context, RRD_ID_LENGTH_MAX, "powersupply.%s", dim);
+
+        prop->st = rrdset_create_localhost(
+                id
+                , ps->name
+                , NULL
+                , dim,
```
Comment (ilyam8) [inline src/collectors/proc.plugin/sys_class_power_supply.c:171]:
```suggestion
                , dim
```

I see this error in ci logs

```c /netdata/src/collectors/proc.plugin/sys_class_power_supply.c: In function 'rrdset_create_simple_prop':
/netdata/src/collectors/proc.plugin/sys_class_power_supply.c:180:9: error: macro "rrdset_create_localhost" passed 13 arguments, but takes just 12
```

Comment (eyusupov) [reply src/collectors/proc.plugin/sys_class_power_supply.c]: Thanks. Not sure how I missed it. :-)

**Comment by eyusupov** [comment]
This is the new screenshot:
image

I was expecting `power` to be showing after `charge` due to the priority `NETDATA_CHART_PRIO_POWER_SUPPLY_POWER`. Am I missing or misunderstanding something? Showing it the way it is is also fine for me, but the code should reflect it then.

File: src/collectors/proc.plugin/sys_class_power_supply.c [hunk 2]
```diff
@@ -116,14 +130,83 @@ static void add_labels_to_power_supply(struct power_supply *ps, RRDSET *st) {
     rrdlabels_add(st->rrdlabels, "device", ps->name, RRDLABEL_SRC_AUTO);
 }
 
+static bool read_simple_property(struct simple_property *prop, bool keep_fds_open) {
+    char buffer[30 + 1];
```
Comment (thiagoftsm) [inline src/collectors/proc.plugin/sys_class_power_supply.c:134]: Please, can you use a `define NAME` to express the meaning of number `30`? Thanks!

File: src/collectors/proc.plugin/sys_class_power_supply.c [hunk 3]
```diff
@@ -116,14 +130,83 @@ static void add_labels_to_power_supply(struct power_supply *ps, RRDSET *st) {
...
+
+    if(unlikely(prop->fd == -1)) {
+        prop->fd = open(prop->filename, O_RDONLY | O_CLOEXEC, 0666);
+        if(unlikely(prop->fd == -1)) {
+            collector_error("Cannot open file '%s'", prop->filename);
+            return false;
+        }
+    }
+
+    if (likely(prop->fd != -1))
```
Comment (thiagoftsm) [inline src/collectors/proc.plugin/sys_class_power_supply.c:144]: I suggest instead to test `prop->fd != -1` again, you write it as an `else` for previous block.

File: src/collectors/proc.plugin/sys_class_power_supply.c [hunk 4]
```diff
@@ -116,14 +130,83 @@ static void add_labels_to_power_supply(struct power_supply *ps, RRDSET *st) {
...
+    {
+        ssize_t r = read(prop->fd, buffer, 30);
```
Comment (thiagoftsm) [inline src/collectors/proc.plugin/sys_class_power_supply.c:146]: When you create a macro (`define`) above, please, do not forget to change here.

File: src/collectors/proc.plugin/sys_class_power_supply.c [hunk 5]
```diff
@@ -116,14 +130,83 @@ static void add_labels_to_power_supply(struct power_supply *ps, RRDSET *st) {
...
+        if(unlikely(!keep_fds_open)) {
+            close(prop->fd);
+            prop->fd = -1;
+        }
+        else if(unlikely(lseek(prop->fd, 0, SEEK_SET) == -1)) {
+            collector_error("Cannot seek in file '%s'", prop->filename);
+            close(prop->fd);
+            prop->fd = -1;
+        }
+    }
```
Comment (thiagoftsm) [inline src/collectors/proc.plugin/sys_class_power_supply.c:167]: Considering that you are having:

```sh
 close(prop->fd);
prop->fd = -1;
```

on three difference places, and on two places you are also calling `collector_error`, probably you can rework this part of the code to have as an inline function.

Comment (eyusupov) [reply src/collectors/proc.plugin/sys_class_power_supply.c:159]: I did not use inline function, but simplified error handling in the way similar to how it's done for multi-dimensional properties

**Comment by eyusupov** [comment]
Screenshot from the latest test run:
image

**Comment by thiagoftsm** [comment]
> I prepared another PR with the same change, where I have unified the code for single and multi-dimensional properties #17346. Let me know if you think it's better or if I should finish that one instead.

Hello,

I am sorry for the delay, I got some days off.
I suggest we finish this first, so you can st the other as draft for a while.

**Comment by thiagoftsm** [comment]
@ilyam8 I cannot see on my host the charts, can you see on yours?

**Review by thiagoftsm — APPROVED** [review]
I do not have the file `power_now` on my environments, but the plugin is running as expected with these changes. I am going to approve waiting for other reviewers that have conditions to see the chart.
Thank you @eyusupov !

**Pull request merged**